### PR TITLE
Add the LV2 plugin format

### PR DESCRIPTION
**Pull request by jpcima**
This an implementation of LV2 format for Surge on Linux.
There was some related discussion at #514.

If @falkTX agrees, I would like to submit it possible upstream integration.
I did not make it with DPF, however I took a few elements of code from his previous work, therefore this contribution is made of mine and his.
(not starting with a framework, I thought I may have more ease to link together VSTGUI and LV2)

The plugin could benefit of a bit of testing, I'll note a few technical items.
- LV2 installs an external description of the plugin itself as a collection of RDF files.
These files needn't be written manually, the plugin will dump them. (see `SurgeLV2Exporter`)
It's integrated in the build as a python script.
- It has the fake Run Loop necessary for X11, which is ticked manually in the idle callback.
- It's a problem if parameter names change in the future, it would break saved parameter states.
If this has to be sbsolutely resolved, parameters must be assigned stable identifiers (symbols).
- About the custom MIDI CC assignment, it cannot set the value of the parameter in the host when it's received, only in the synth. LV2 is explicitly disallowed to change control input values. (that's a feature of LV2, or a restriction, depending who you ask)
- I don't know if I get the `ppqPos` thing right. I would like if someone can verify it.

To do:
- [x] ~~allow LV2 to build on all OS~~ (do in the nightlies following release 1.6.2)
- [X] copy lv2 files into the debian package
- [ ] resize handling

### CODE REVIEW

File: premake5.lua [hunk 1]
```diff
@@ -659,3 +659,57 @@ if (os.istarget("macosx")) then
     end
 
 end
+
+-- VST3 PLUGIN --
```
Comment (baconpaul) [inline premake5.lua]: Can you adjust this comment? Thank you!

**Comment by baconpaul** [comment]
Thank you very much for both doing and submitting this. I'm sure the entire community will appreciate it.

As you may notice, though: the commit broke our CI pipeline by breaking the windows builds both 64 and 32 bit. I'm sure this is a misplaced ifdef or include. It looks like the windows build is including the lv2 through SurgeSynthesizer.cpp. If you could fix that up then I'll find time this week to do a review and stuff. I'll also skim the code now and add a few inline comments as I see them.

**Comment by baconpaul** [comment]
Oh and on parameter names / parameter IDs - that is a problem in AU VST2 and VST3 land also. Over on slack I was chatting about doing an experimental branch where I tried to fix that so we can easily add parameters all hosts. But we want to do that post 1.6.2.

File: premake5.lua (hunk 1 above)
Comment (jpcima) [reply premake5.lua]: You are right, I've missed it totally. Fixed

**Comment by baconpaul** [comment]
Oh and I think the reason it is breaking windows is because it is trying to build windows LV2! I don't think we need that.

If you take a lookin premake you can see we wrap the audio unit in an if macOS. I think do the same for the LV2 block and we should pass CI.

One other thing. In `installer_linux` is a script which assembles components into a .deb file. We probably want to add the .lv2 there also?

Finally: I wonder if your moving the creation of that timer will fix our linux bitwig problems with VST3? That would be a cool side effect!

**Comment by jpcima** [comment]
>Oh and I think the reason it is breaking windows is because it is trying to build windows LV2! I don't think we need that.

It's true, at least it's the only reason I can see.
One way to resolve it is to check out the lv2 header set into surge.

>One other thing. In installer_linux is a script which assembles components into a .deb file. We probably want to add the .lv2 there also?

I will add a task check list in the top post.

>Finally: I wonder if your moving the creation of that timer will fix our linux bitwig problems with VST3? That would be a cool side effect!

It's been necessary to move the timers after creating the runloop, or an error would be produced.

**Comment by baconpaul** [comment]
> > Oh and I think the reason it is breaking windows is because it is trying to build windows LV2! I don't think we need that.
> It's true, at least it's the only reason I can see. One way to resolve it is to check out the lv2 header set into surge.

An easier way is to just include the entire lv2 section in an

```
if (os.istarget("linux")) then
```

block. I think that's more appropriate. 

> It's been necessary to move the timers after creating the runloop, or an error would be produced.

Yeah! Our VST3 works in reaper but not bitwig because of some error. I wonder if it is the error that made you move this. (In VST3 I do a similar-but-different runloop thing to your lv2 and falks vst2)

**Comment by jpcima** [comment]
>An easier way is to just include the entire lv2 section in an

LV2 is not linux only, it may be able to work for all.
But, I didn't test these other OSes, so a conditional will be wise for the moment.

I'll be back to implement this in a moment, or, feel free to do so yourself if you want.

**Comment by mortfell** [comment]
Oh cool!
I have Ardour installed on Mac so I can do some testing of LV2 on that platform.

**Comment by baconpaul** [comment]
Cool; I'm thinking for 1.6.2 (which @jpcima we are trying to release in the next couple of weeks) we should do linux only and then once we've shipped that perhaps turn on the mac lv2 in the nightlies? Just out of the principle of 'least change' really as we are stabilizing. I would like to try and merge this change before 1.6.2 though for linux!

Although LV2 works on all platforms it seems the demand really is from the linux community right now. Am I wrong about that?

**Comment by mortfell** [comment]
that seems wise to me as well.

**Comment by falkTX** [comment]
LV2 has grown a lot in Linux due to VST having serious licensing issues.
It got better with VST3, but a little too late. Plus VST3 is a serious PITA, I am yet to see someone that actually likes to work with it... contrary to LV2, where we get community involvement quite often, and even making their own unofficial extensions (like Ardour inline-displays in its mixer, and MOD with its remote modgui in a web browser)

Nothing stops Windows and macOS (or any commercial DAW) to support LV2.
But since commercial DAWs treat Linux as non-priority, formats that matter to Linux end up in the same position.
Pretty much all opensource DAWs support LV2 though :)
So you get stuff like Audacity, Ardour, Qtractor, MusE, Traverso and smaller hosts like Carla and Ingen supporting LV2, but not commercial ones.

**Comment by baconpaul** [comment]
Yeah my desire to hold on windows and mac is just "we are 3 weeks from and in the stabilization period around a release"; not "we shouldn't do it".

And, speaking as the person who has fixed most of the VST3 surge problems, you are still yet to see someone who actually likes working with it. :)

File: src/lv2/SurgeLv2Ui.cpp [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+SurgeLv2Ui::~SurgeLv2Ui()
+{
+   _instance->setEditor(nullptr);
+   _editor->close();
+}
+
+void SurgeLv2Ui::setParameterAutomated(int externalparam, float value)
+{
+   _writeFn(_controller, externalparam, sizeof(float), 0, &value);
```
Comment (falkTX) [inline src/lv2/SurgeLv2Ui.cpp]: you need to offset the audio and event ports here, no?

Comment (jpcima) [reply src/lv2/SurgeLv2Ui.cpp]: I have order the control ports first on purpose, in order to avoid dealing with any offsets.

**Comment by baconpaul** [comment]
OK this is great. I'm in the middle of fixing up another bug or two and would like to try a build of this myself before I merge it, but at least in the place where we intersect with non-lv2 code this looks good to me.

@falkTX I see you've been looking too. Does it look appropriate to merge to you? You know so much more about this format than I do! Thanks for reviewing!

And @jpcima again - thanks for doing this. Really great.

File: src/lv2/SurgeLv2Ui.cpp (hunk 2 above)
Comment (falkTX) [reply src/lv2/SurgeLv2Ui.cpp]: ah, smart. :)

**Comment by falkTX** [comment]
From a quick overview it seems fine to me.
There are a few TODO items, but things that can be improved later on perhaps.
(like having no tempo sync)

I worry about the lack of resize reporting to the host though.
@jpcima there is a way to scale the UI, which should trigger a resize of the host window.
did you try that?

**Comment by falkTX** [comment]
hmm I might be confusing things, but seems to me UI idle and runloop idle are not in the same thread.
if that is the case, this will cause issues.

**Comment by mortfell** [comment]
I have another desktop running Linux here. 
I don't know anything about LV2 development, but I'd be happy to fire up that machine and spend a few hours testing LV2 and report any bugs I find on Ardour/Renoise once it's building correctly!

**Comment by jpcima** [comment]
>hmm I might be confusing things, but seems to me UI idle and runloop idle are not in the same thread. if that is the case, this will cause issues.

You can find the statement `self->_runLoop->execIdle()` inside `SurgeLv2Ui::uiIdle`.

>I worry about the lack of resize reporting to the host though.

Does it matter when the window is a fixed-size? I'll examine it later.


There was another thing, it's the parameters all defined in the domain 0-1.
I was unsure if I should keep them in 0-1 or denormalized. When intermediate conversions are made, it's not guaranteed to get an equal value back after doing a sequence of `set` and a `get`.
But maybe it's not a concern any longer, after I've implemented a parameter strategy which similar to that used in DPF, so I could reconsider that part also.

**Comment by falkTX** [comment]
> > I worry about the lack of resize reporting to the host though.
> 
> Does it matter when the window is a fixed-size? I'll examine it later.

the window is not truly fixed size. it is non-user resizable, but programmatically it can scale which changes its size.

**Comment by baconpaul** [comment]
> > Does it matter when the window is a fixed-size? I'll examine it later.
> 
> the window is not truly fixed size. it is non-user resizable, but programmatically it can scale which changes its size.

You would see this by using the "menu / zoom / zoom to ___" feature in the current synth.

> From a quick overview it seems fine to me.
> There are a few TODO items, but things that can be improved later on perhaps.
> (like having no tempo sync)

Thanks. I'm happy to merge it in current state, then, if you agree it's ready @jpcima, once I've done a test quickly on linux mac and win of the vst3 (which had some overlapping minor changes).

Let me know if code as stands is at a good merge-and-make-issues point.

**Comment by baconpaul** [comment]
Oh and also @jpcima I'll squash this PR into one commit for you if you want; but if you would rather do that yourself and write the merged commit message, feel free to rebase and force push. Just lemme know!

**Comment by baconpaul** [comment]
Ahh OK so this isn't quite ready to merge I don't think, mostly because of build configuration stuff.

Turns out the CI build failed, but failed silently. I'll try and figure out why exactly build-linux is suppressing errors (#1095) but if you look here

https://dev.azure.com/surge-synthesizer/surge/_build/results?buildId=1792

you'll see the lv2 build failed. When I try it manually on my ubuntu 18.04 I get the same message as CI:

```
In file included from src/lv2/SurgeLv2Wrapper.h:3:0,
                 from src/lv2/SurgeLv2Export.cpp:1:
src/lv2/AllLv2.h:2:10: fatal error: lv2/core/lv2.h: No such file or directory
 #include <lv2/core/lv2.h>
          ^~~~~~~~~~~~~~~~
compilation terminated.
```

I did an `apt-get install lv2-dev` and that gave me some files but not that file.

So where are you getting the lv2/core/lv2.h file from in your system and how do we make it available to our linux build?

Thanks!

**Comment by baconpaul** [comment]
OK I've fixed our pipelines.

If you can rebase to master or at least cherry pick in https://github.com/surge-synthesizer/surge/commit/c931452ac38e6d0c2f9b9da7795d5e216c019fc8 you will see the failure in CI for linux lv2.

We just need to figure out where to get those libs in our build environment. Lots of options

1. If there's a well defined package in apt-get we can add that to the pipeline
2. If there's a github repo we can add it as a submodule in libs/ and you can point to that
3. If there's neither we should have a chat here on the issue.

Thanks!

**Comment by trebmuh** [comment]
https://packages.debian.org/search?mode=path&suite=buster&section=all&arch=any&searchon=contents&keywords=lv2.h

**Comment by jpcima** [comment]
I rebased on master. For a generic solution, we may include David Robillard's lv2 repository as submodule.
https://github.com/drobilla/lv2

I think, the resizing issue is not necessary to fix right away, but I'll look.

However, the natural valuation of parameters must be possible, if I'm not mistaken, so I'll give this a try before finalizing the lv2 parameters. I let you know if that works.

**Comment by baconpaul** [comment]
The problem with parameters in surge is that their scale changes dynamically based on type. So for instance the 3rd OSC parameter can be dB in one oscillator and FM feedback in another; with totally different ranges. The get/set_parameter_f01 gives you a normalized 0/1 value. This can be, I agree, somewhat unsatisfactory from an automation UI perspective 

And happy for you to add that repo as a submodule with your PR and then whack the premake to point at it. We've put (most of) our submodules in libs and I'd like to do that here. 

And cool on the rebase. As you can see, the CI now fails properly! 

Let me know if you need help with any of this. Appreciate you doing this!

**Comment by jpcima** [comment]
Alright macOS is trying to include the whole `libs/**` recursively, it's why I have failure apparently.

**Comment by baconpaul** [comment]
I’m at my mac now. Lemme see if I can fix it for you.

**Comment by baconpaul** [comment]
Wow I'm amazed it ever worked even before your change! Getting it sorted now.

**Comment by baconpaul** [comment]
[premake_patch.txt](https://github.com/surge-synthesizer/surge/files/3571067/premake_patch.txt)

Hi

If you apply that patch to your branch and re-push it should all work.

Sorry that was such a mess. We hadn't really touched it since inception and we just got lucky that it didn't break. Definitely more explicit now.

**Comment by baconpaul** [comment]
OK cool - CI built correctly. Let me just give it a whirl on linux then I will squash it down and merge it.

**Comment by jpcima** [comment]
Alright, either I figure the resizing and implement it in this PR, or it waits a next one.

**Comment by baconpaul** [comment]
Wahey the LV2 works in carla!! AWESOME!

I propose get this PR done without resize and then figure it separately.

There is one last tiny change if you don't mind. In src/common/gui/CAboutBox.cpp around like 50 you see a cascade of #if TARGET_ to draw the flavor on the about screen. If you can add LV2 to that then rebase to a single commit, add the final commit comment, and force push up the single commit, I'll happily merge it.

If you'd rather tackle zoom I'll wait too. Just let me know.

**Comment by baconpaul** [comment]
Or if you want I can just squash this now and add the about box myself. Your choice!

**Comment by jpcima** [comment]
Added the note in the about dialog. For resize, it can wait, because I will take needed time to understand the mechanics of it, as I've not used before.

For squashing, I'm sure Github offers the option under the merging button.

**Comment by baconpaul** [comment]
Yup. That's the button I always press. Just making sure you don't mind me rewriting the commit message.

Squashing and merging now.

Thank you very much for this contribution!

**Comment by baconpaul** [comment]
Merged!

**Comment by jpcima** [comment]
Thanks @baconpaul !
I have to say, it's amazing your enthusiasm towards Linux contributions, and it's a pleasure contributing. Thank you a lot really.

**Comment by baconpaul** [comment]
Thanks for the kind words! I've been programming unixen since before linux so it's easy to support, especially since most of the hard platform-specific work has been done by you and @falkTX and other contributors. It's all good stuff!

If you want to work on any other surge issues also, I'd be happy to point you at a few good low hanging fruit for after 1.6.2 ships also. (Like I said: we are really trying to stabilize so we can ship 1.6.2 at end of Sept).

**Comment by jpcima** [comment]
Well yes, I don't mind to work on fixing a thing or two when it's within my ability and available time.
It's already kinda what I do in the LibraZiK distribution of music software; I go to fix software problems left and right based on user reports.

It's in this very context, that I've had in my work list to implement the Surge LV2.
Being free and open-source software distributions, we have trouble offering VST in official package repositories, which are non-free due to VST SDK license.

From this point, distributing in a Linux distribution like Debian is no longer blocking.

**Comment by baconpaul** [comment]
Cool. Well here's the linux list: https://github.com/surge-synthesizer/surge/issues?q=is%3Aopen+is%3Aissue+label%3Alinux 
but most of those are vstgui problems

let me know where your interest is and I'm happy to help!

**Comment by baconpaul** [comment]
Oh and the linux VST3 is 100% clean GPL code. no problems there, since vst3sdk is dual license. So we could distribute the VST3 and LV2 in any free software setting. Just not VST2

**Comment by jpcima** [comment]
Linux hosting support of VST3 is not widespread.
I even think, as far as only free software are considered, it doesn't exist at all.

It's possible the LibraZiK distro is going to push some testing packages pretty soon, and there will be surely some tester feedback out of it, so I'll keep you in touch.

**Comment by baconpaul** [comment]
Oh yeah. The state of play of Linux VST3 is pretty grim.  The Steinberg examples don't work in some cases even!

But our VST3 does work in reaper which seems like a very popular, albeit non-free, linux environment.

My point was more: If you are distrubuting free-only software you could also include the VST3 and keep being free-only.

**Comment by mortfell** [comment]
Hello to everyone doing LV2 tuff for Surge!
I installed a recent linux distro on an older computer and did a bunch of testing with Surge on different platforms tonight. Did a handful of bug reports and pooled some info in: #1105 

I am pretty experienced as a Linux user, it was my main operating system for a few years, and I did lots of audio stuff, but it's been a while.

If there is anything at all I could do to give better info about my setup please let me know!

**Pull request merged**